File: python/problem_66.py
class Solution(object):
    def plusOne(self, digits):
        """
        :type digits: List[int]
        :rtype: List[int]
        """

        # iterate from the rightmost digit and check if it's 9
        # if it isn't, increment by one
        # if it isn't continue and carry over the increment

        n = len(digits)

        # Iterate from the rightmost digit to the leftmost
        for i in range(n - 1, -1, -1):
            if digits[i] < 9:
                digits[i] += 1
                return digits  # No carry-over, so return the modified list
            else:
                digits[i] = 0  # Current digit is 9, set to 0 and carry over

        # If all digits were 9s (e.g., [9,9,9]), a carry-over remains
        # Prepend a 1 to the list
        return [1] + digits

File: java/leetcode_problems/problem_66.java
package leetcode_problems;

public class problem_66 {
    public int[] plusOne(int[] digits) {
        int n = digits.length;

        // Iterate from the rightmost digit (least significant)
        for (int i = n - 1; i >= 0; i--) {
            // If the current digit is less than 9, increment it and return the array
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            } else {
                // If the current digit is 9, set it to 0 and continue to the next digit (carry-over)
                digits[i] = 0;
            }
        }

        // If the loop completes, it means all digits were 9 (e.g., [9,9,9])
        // In this case, we need a new array with an additional digit at the beginning
        int[] newDigits = new int[n + 1];
        newDigits[0] = 1; // The new leading digit is 1 (e.g., [1,0,0,0])
        return newDigits;
    }

    public static void main(String[] args) {
        int[] digits = {9,4,6,7,8};
        problem_66 object = new problem_66();
        int[] arr = object.plusOne(digits);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }

    }
}